services/federator/src/Federator/Monitor/Internal.hs

```diff
@@ -0,0 +1,355 @@
+import Polysemy (Embed, Member, Members, Sem, embed)
+import qualified Polysemy
+import qualified Polysemy.Error as Polysemy
+import qualified Polysemy.Resource as Polysemy
+import Polysemy.TinyLog (TinyLog)
+import qualified Polysemy.TinyLog as Log
+import System.FilePath (dropTrailingPathSeparator, splitFileName)
+import System.INotify
+import System.Logger (Logger)
+import qualified System.Logger.Message as Log
+import System.Posix.ByteString (RawFilePath)
+import System.X509
+import Wire.API.Arbitrary
+
+data Monitor = Monitor
+  { monINotify :: INotify,
+    monTLS :: IORef TLSSettings,
+    monWatches :: IORef Watches,
+    monSettings :: RunSettings,
+    monHandler :: WatchedPath -> Event -> IO (),
+    monLock :: MVar ()
+  }
+
+-- This is needed because the normal Posix file system API uses strings, while
+-- the inotify API uses bytestrings.
+-- /Note/: File paths are strings obtained using the "file system encoding",
+-- which is the same as the locale encoding, but uses some escaping tricks to
+-- be able to represent arbitrary data as strings.
+rawPath :: FilePath -> IO RawFilePath
+rawPath path = do
+  encoding <- getFileSystemEncoding
+  withCStringLen encoding path packCStringLen
+
+data WatchedPath
+  = WatchedFile RawFilePath
+  | WatchedDir RawFilePath (Set RawFilePath)
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving (Arbitrary) via (GenericUniform WatchedPath)
+
+mergePaths :: [WatchedPath] -> (Set WatchedPath)
+mergePaths = Set.fromList . merge . sort
+  where
+    merge [] = []
+    merge [w] = [w]
+    merge (w1 : w2 : ws) = case (w1, w2) of
+      (_, WatchedFile _) -> w1 : w2 : merge ws
+      (WatchedDir dir1 paths1, WatchedDir dir2 paths2)
+        | dir1 == dir2 -> merge (WatchedDir dir1 (paths1 <> paths2) : ws)
+      _ -> w1 : merge (w2 : ws)
+
+watchedPath :: WatchedPath -> RawFilePath
+watchedPath (WatchedFile path) = path
+watchedPath (WatchedDir dir _) = dir
+
+watchPathEvents :: WatchedPath -> [EventVariety]
+watchPathEvents (WatchedFile _) = [CloseWrite]
+watchPathEvents (WatchedDir _ _) = [MoveIn, Create]
+
+-- Since we are watching a filesystem path, and not an inode, we need to replace a
+-- file watch when the file gets overwritten.
+-- This type is a map of paths to watches used to keep track of both file and
+-- directory watches as they get deleted and recreated.
+type Watches = Map RawFilePath (WatchDescriptor, WatchedPath)
+
+runSemDefault :: Logger -> Sem '[TinyLog, Embed IO] a -> IO a
+runSemDefault logger = Polysemy.runM . Log.runTinyLog logger
+
+logErrors ::
+  Members '[TinyLog, Polysemy.Error FederationSetupError] r =>
+  Sem r a ->
+  Sem r a
+logErrors action = Polysemy.catch action $ \err -> do
+  Log.err $
+    Log.msg ("federation setup error while updating certificates" :: Text)
+      . Log.field "error" (showFederationSetupError err)
+  Polysemy.throw err
+
+logAndIgnoreErrors ::
+  Member TinyLog r =>
+  Sem (Polysemy.Error FederationSetupError ': r) () ->
```

Comment:
Can this be only of the unit return type or can it be any `a` in `Sem (Polysemy.Error FederationSetupError ': r) ()`?

Reply:
Since errors are ignored, it can only be `()`.